[Feature] Type inference for variables

mohammadfawaz opened the pull request:
## Motivation

Closes #28630

The goal is to make it easy to define variables so that something like this would work:
```
let x = 1u32;
let y = Foo { .. };
// .. etc
```
where the types of `x` and `y` are automatically set by the compiler.

- `let` type annotation is now optional
- Continue to error out on annotations that don't match the type of the RHS
- Infer the types of defined variables from the type of the RHS.
- Continue to handle tuples and their errors as before. This error is no longer used though:
```rust
@formatted
    invalid_tuple {
        args: (),
        msg: "Tuples must be explicitly typed in Leo".to_string(),
        help: Some("The function definition must match the function return statement".to_string()),
    }
```

## Test Plan
- A few tests now emit the intended errors. How to slightly modify a few of them.
- Added two new tests

d0cd commented on compiler/parser/src/parser/statement.rs:
```diff
@@ -250,8 +250,9 @@ impl<N: Network> ParserContext<'_, N> {
-        self.expect(&Token::Colon)?;
-        let type_ = self.parse_type()?.0;
+
+        // The type annotation is optional
```
```suggestion
        // The type annotation is optional.
```

d0cd commented on tests/expectations/parser-statement/statement/conditional_fail.out:
```diff
@@ -1,5 +1,5 @@
+Error [EPAR0370017]: Could not parse the implicit value: 2.
+    --> test_0:1:25
      |
    1 | if true {} else let x = 2;
```
Should inference also take care of numeric literals?

mohammadfawaz replied:
Yes. I already opened this issue https://github.com/ProvableHQ/leo/issues/28626 which I'm working on next. All this PR is doing is type inference for variables.

d0cd commented on tests/tests/compiler/type_inference/basic_inference.leo:
```diff
@@ -0,0 +1,29 @@
+    transition main(x: u32, y: u32, addr: address, arr: [u32; 4]) 
+        -> (S, S, u64, u32, u32, bool, u32, u32) {
+        let s1 = S { x, y };
+        let s2: S = s1;
```
Just curious, does inference work if we remove this type annotation?

mohammadfawaz replied:
Yes. The next line is `let s3 = s1;` which is the same but without the annotation and it works fine!

mohammadfawaz commented on tests/tests/compiler/function/shadow_parameter_fail.leo:
```diff
@@ -1,10 +1,10 @@
 
 program test.aleo {
-    function tester(hi: u8) -> u8 {
+    function transition(hi: u8) -> u8 {
```
Ugh I meant to change `function` to `transition` here because programs must have at least one transition. I changed the function names instead 🤦. Fixing.

mohammadfawaz commented on tests/tests/compiler/multibyte/multibyte_fail.leo:
```diff
@@ -27,7 +27,7 @@ program bubblesort.aleo {
                // Move the smaller elements forward
                if arr[j+1u8] < arr[j] {
-                   let swap = arr[j];
+                   let swap: = arr[j];
```
Yeah the point of the test is to check that errors point to the right spot in the file despite the code containing multibyte characters. Now that the type annotation error is gone, I had to introduce a new one.

d0cd reviewed (commented):
The implementation looks good. Just some minor nits.
I'm pleasantly surprised this was all that was required to implement inference for variables!

To confirm my understanding, the reason why we don't need to reach for heavyweight machinery like Algorithm W is because our language is simple (no parametric polymorphism, first-class functions, etc.) so local inference is sufficicent?

mohammadfawaz commented on tests/tests/compiler/type_inference/bad_tuples.leo:
```diff
@@ -0,0 +1,13 @@
+program foo.aleo {
```
Fixed.

mohammadfawaz commented on compiler/passes/src/type_checking/expression.rs:
```diff
@@ -1038,30 +1038,55 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
+            if let Type::Tuple(expected_types) = expected {
+                // If the expected type is a tuple, then ensure it's compatible with `input`
+
+                // Firt, make sure that the number of tuple elements is correct
```
Fixed.

mohammadfawaz commented on compiler/passes/src/type_checking/expression.rs:
```diff
@@ -1038,30 +1038,55 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
+            } else {
+                // If the expected type is not a tuple, then we just error out
+
+                // This is the type of the tuple based on its individual fields
```
Fixed.

mohammadfawaz commented:
> Cool, looks great. I made a couple tiny remarks. The only other thing I have some concern about is Futures. I imagine there's not actually anything that could go wrong, but it'd be good to make sure the "type inference" we do around futures doesn't interact with this type inference in any problematic way. I think ideally I'd like to see a test in which `async function foo(fut: Future)` is called by two different async transitions, passing in two different futures, which have been declared without a type: `let fut = some_program.aleo/make()` or whatever. Let me know if that doesn't make sense to you.

Added a passing test and a failing test that use this pattern. In fact, this uncovered a small inference issue in `visit_call` that I had overlooked and just fixed. Thanks!

mohammadfawaz commented:
> The implementation looks good. Just some minor nits. I'm pleasantly surprised this was all that was required to implement inference for variables!
> 
> To confirm my understanding, the reason why we don't need to reach for heavyweight machinery like Algorithm W is because our language is simple (no parametric polymorphism, first-class functions, etc.) so local inference is sufficicent?

Yeah exactly. The key point is that type inference for Leo does not require any changes to the AST itself. It's all done in the map that keeps track of the types of nodes. For more complex languages that have generics for example, the AST would have to be modified (making copies of generic functions for example, trait constraint solving, etc.) and that's when all the complex algorithms are required.
